rfcs/0000-cosign-integration.md

```diff
@@ -0,0 +1,162 @@
+  - `cosign copy` to copy images along with their signatures from one registry
+    to another.
+- Sign container images by tag or by digest.
+- Verify container image signatures by specifying either a tag or a digest.
+- Verify container image signatures after relocation.
+
+Moreover, `cosign` is a project that has had a lot of activity and community
+involvement recently.
+
+## Outcome
+
+`kpack` integrates with `cosign` to sign images it builds and push the
+signatures to a registry so that users can ensure the chain of custody of a
+generated artifact.
+
+This proposal aims to cover specifically the flow of signing an image produced
+by `kpack Image` resource without verification of base or builder images pulled
+in the process.
+
+## Actions to take
+
+### Enable image signing using `cosign`
+
+- Implement a flow that generates the signature payload for the image, then
+  calculates its signature and pushes it either to the registry where the image
+  is located, using the same credentials that were used to push the image, or
+  [to the registry specified in the `COSIGN_REPOSITORY` environment variable](#key-generation-and-storage).
+  This flow must happen after the image has been pushed to the registry.
+
+- If `cosign` fails to sign an image, the build should fail and output an error
+  message in the build log, so the operator can troubleshoot the issue. The
+  error messages should also be added in any other places where error messages
+  are presented.
+
+- Whenever `kpack` signs an image using `cosign`, it should add these
+  annotations:
+  - Build number.
+  - Build time.
+
+- Signing with `cosign` should not affect any configurations that enable signing
+  using the Notary v1 mechanism. These two signing mechanisms should be able to
+  coexist.
+
+### Key generation and storage
+
+`cosign` can use file-based keys, key management systems or hardware keys to
+ingest private keys, as well as a keyless flow that requires integration with
+[Rekor](https://github.com/sigstore/rekor).
+
+`kpack` should not generate the keys used for signing and verification. The user
+should pass them in using one of the mechanisms supported by `cosign`.
+In this RFC, we suggest that the first mechanism implemented be the use of
+Kubernetes `Secret`s for sending the relevant data into the build pod, by means
+of attaching secrets to the service account used by `kpack`.
+These secrets should follow the format generated by
+[`cosign generate-key-pair -k8s <namespace>/<name>`](https://github.com/sigstore/cosign/pull/345).
+
+The service account can have more than one `cosign` secret attached to it. For
+each `cosign` secret attached, a separate signature should be generated and
+stored.
+
+`kpack` will be able to determine that the user wants to sign images using
+`cosign` by scanning the annotations in each of the secrets. Secrets to be used
+by `cosign` signing should be annotated with `kpack.io/cosign-credentials`.
```

Comment:
Is the `kpack.io/cosign-credentials` annotation necessary can kpack just detect the `cosign.*` keys?

Reply:
That would work, too. I originally thought of adding the specific annotation to prevent accidental signing, as a mean of making intention clear.

Reply:
I am a bit concerned that would require creating secrets with cosign and then altering them to support use in kpack. Would you object to being permissive and signing with any "cosign" secret?

Reply:
I see. That's fair, I can update it to remove the requirement of having an empty annotation here. I also just recalled that my original rationale was that the user would need to manually update the secret in any case, in order to add the `COSIGN_REPOSITORY` and `COSIGN_DOCKER_MEDIA_TYPES` variables, but this is only true when the signature is being pushed to a different repository or legacy media types compatibility is required.